PR: https://github.com/odoo/odoo/pull/176212

From: 7d0e4aaccbeef166317bd7180108372be3dde39d
From: Einar Andersen (eian)
Date: 2024-09-17 08:32:11

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = ['mail.thread']

Total Changes: 98

[IMP] hr: UX and onboarding improvements

A rework of the UX/UI for parts of the HR/Employee app which includes:
* Skills are now translatable, enabling the detection of skills on resumes written in languages other than English.
* Improved order on fields in employee form view.
* Adding/improving placeholders for multiple fields.
* Adding/improving/removing tooltips for multiple fields.
* Adding/improving No Content helpers for multiple views to make the intention of the view clearer.
* Adding and removing measurement fields for graph and pivot views to improve usability.
* Spelling improvements.
* Improved wording for multiple field strings.
* Hiding buttons on employee form when employee is archived.
* Creating activities on departments is now possible.
* On/Off-boarding: If the employee/manager/coach is not linked to a user we now go up the chain of management until we find a manager linked to a user. This should allow a more seamless user experience.
* Adding filter for employees in Contact app.
* A new tab "Payroll" has been added in the Employee form view to gather all the payroll information.
* A new field "Legal Name" has been added to accommodate employees whose legal names diverge from their preferred name.
* Replacing the "Name" field with the "Legal Name" field in various payroll and contract situations.

task-3930677

closes odoo/odoo#176212

Related: odoo/enterprise#68169
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/hr_department.py
+++ b/addons/hr/models/hr_department.py
@@ -8,7 +8,7 @@ from odoo.exceptions import ValidationError
 class Department(models.Model):
     _name = "hr.department"
     _description = "Department"
-    _inherit = ['mail.thread']
+    _inherit = ['mail.thread', 'mail.activity.mixin']
     _order = "name"
     _rec_name = 'complete_name'
     _parent_store = True

--- a/addons/hr/models/hr_employee.py
+++ b/addons/hr/models/hr_employee.py
@@ -90,7 +90,7 @@ class HrEmployeePrivate(models.Model):
     identification_id = fields.Char(string='Identification No', groups="hr.group_hr_user", tracking=True)
     passport_id = fields.Char('Passport No', groups="hr.group_hr_user", tracking=True)
     bank_account_id = fields.Many2one(
-        'res.partner.bank', 'Bank Account Number',
+        'res.partner.bank', 'Bank Account',
         domain="[('partner_id', '=', work_contact_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]",
         groups="hr.group_hr_user",
         tracking=True,
@@ -128,7 +128,7 @@ class HrEmployeePrivate(models.Model):
             ('contractor', 'Contractor'),
             ('freelance', 'Freelancer'),
         ], string='Employee Type', default='employee', required=True, groups="hr.group_hr_user",
-        help="The employee type. Although the primary purpose may seem to categorize employees, this field has also an impact in the Contract History. Only Employee type is supposed to be under contract and will have a Contract History.")
+        help="Categorize your Employees by type. This field also has an impact on contracts. Only Employees, Students and Trainee will have contract history.")
 
     job_id = fields.Many2one(tracking=True)
     # employee in company

--- a/addons/hr/models/hr_employee_base.py
+++ b/addons/hr/models/hr_employee_base.py
@@ -42,7 +42,7 @@ class HrEmployeeBase(models.AbstractModel):
         ("home", "Home"),
         ("office", "Office"),
         ("other", "Other")], compute="_compute_work_location_name_type")
-    user_id = fields.Many2one('res.users')
+    user_id = fields.Many2one('res.users', help="")
     share = fields.Boolean(related='user_id.share')
     resource_id = fields.Many2one('resource.resource')
     resource_calendar_id = fields.Many2one('resource.calendar', check_company=True)

--- a/addons/hr/models/ir_ui_menu.py
+++ b/addons/hr/models/ir_ui_menu.py
@@ -16,6 +16,5 @@ class IrUiMenu(models.Model):
                 ('manager_id', 'in', self.env.user.employee_ids.ids)
             ]))
             if not is_department_manager:
-                res.append(self.env.ref('hr.menu_hr_department_tree').id)
                 res.append(self.env.ref('hr.menu_hr_department_kanban').id)
         return res

--- a/addons/hr/models/mail_activity_plan_template.py
+++ b/addons/hr/models/mail_activity_plan_template.py
@@ -21,29 +21,82 @@ class MailActivityPLanTemplate(models.Model):
             if template.responsible_type in {'coach', 'manager', 'employee'}:
                 raise ValidationError(_('Those responsible types are limited to Employee plans.'))
 
+    def _get_closest_parent_user(self, employee, responsible, error_message):
+        responsible_parent = responsible
+        viewed_responsible = [employee]
+        while True:
+            if not responsible_parent:
+                return {
+                    'error': error_message,
+                    'responsible': False
+                }
+            if responsible_parent.user_id:
+                return {
+                    'error': False,
+                    'responsible': responsible_parent.user_id
+                }
+            if responsible_parent in viewed_responsible:
+                return {
+                    "error": _(
+                        "Oops! It seems there is a problem with your team structure.\
+                        We found a circular reporting loop and no one in that loop is linked to a user.\
+                        Please double-check that everyone reports to the correct manager."
+                    ),
+                    "responsible": False,
+                }
+            else:
+                viewed_responsible.append(responsible_parent)
+                responsible_parent = responsible_parent.parent_id
+
     def _determine_responsible(self, on_demand_responsible, employee):
         if self.plan_id.res_model != 'hr.employee' or self.responsible_type not in {'coach', 'manager', 'employee'}:
             return super()._determine_responsible(on_demand_responsible, employee)
-        error = False
-        responsible = False
+        result = {"error": "", "responsible": False}
         if self.responsible_type == 'coach':
             if not employee.coach_id:
-                error = _('Coach of employee %s is not set.', employee.name)
-            responsible = employee.coach_id.user_id
-            if employee.coach_id and not responsible:
-                error = _("The user of %s's coach is not set.", employee.name)
+                result['error'] = _('Coach of employee %s is not set.', employee.name)
+            result['responsible'] = employee.coach_id.user_id
+            if employee.coach_id and not result['responsible']:
+                # If a plan cannot be launched due to the coach not being linked to an user,
+                # attempt to assign it to the coach's manager user. If that manager is also not linked
+                # to an user, continue searching upwards until a manager with a linked user is found.
+                result = self._get_closest_parent_user(
+                    employee=employee,
+                    responsible=employee.coach_id.parent_id,
+                    error_message=_(
+                        "The user of %s's coach is not set.", employee.name
+                    ),
+                )
+
         elif self.responsible_type == 'manager':
             if not employee.parent_id:
-                error = _('Manager of employee %s is not set.', employee.name)
-            responsible = employee.parent_id.user_id
-            if employee.parent_id and not responsible:
-                error = _("The manager of %s should be linked to a user.", employee.name)
+                result['error'] = _('Manager of employee %s is not set.', employee.name)
+            result['responsible'] = employee.parent_id.user_id
+            if employee.parent_id and not result['responsible']:
+                # If a plan cannot be launched due to the manager not being linked to an user,
+                # attempt to assign it to the manager's manager user. If that manager is also not linked
+                # to an user, continue searching upwards until a manager with a linked user is found.
+                result = self._get_closest_parent_user(
+                    employee=employee,
+                    responsible=employee.parent_id.parent_id,
+                    error_message=_(
+                        "The manager of %s should be linked to a user.", employee.name
+                    ),
+                )
+
         elif self.responsible_type == 'employee':
-            responsible = employee.user_id
-            if not responsible:
-                error = _('The employee %s should be linked to a user.', employee.name)
-        if error or responsible:
-            return {
-                'responsible': responsible,
-                'error': error,
-            }
+            result['responsible'] = employee.user_id
+            if not result['responsible']:
+                # If a plan cannot be launched due to the employee not being linked to an user,
+                # attempt to assign it to the manager's user. If the manager is also not linked
+                # to an user, continue searching upwards until a manager with a linked user is found.
+                result = self._get_closest_parent_user(
+                    employee=employee,
+                    responsible=employee.parent_id,
+                    error_message=_(
+                        "The employee %s should be linked to a user.", employee.name
+                    ),
+                )
+
+        if result['error'] or result['responsible']:
+            return result
